pkg/backup/backup_compaction.go

```diff
@@ -45,8 +48,81 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/cockroachdb/errors"
 	"github.com/gogo/protobuf/types"
+	"github.com/lib/pq"
 )
 
+var (
+	backupCompactionThreshold = settings.RegisterIntSetting(
+		settings.ApplicationLevel,
+		"backup.compaction.threshold",
+		"the maximum length of a backup chain before compaction occurs (0 to disable)",
+		0,
+		settings.WithVisibility(settings.Reserved),
+	)
+
+	backupCompactionWindow = settings.RegisterIntSetting(
+		settings.ApplicationLevel,
+		"backup.compaction.window",
+		"the number of backups to compact per compaction (must be greater than one and less than threshold)",
+		3,
+		settings.WithVisibility(settings.Reserved),
+	)
+
+	scheduleOnly = true // Test knob to only compact backups on a schedule.
+)
+
+// maybeCompactBackups will conditionally run a compaction job after a backup
+// job if the backup chain exceeds the compaction threshold. It is passed
+// the job details of the backup job that just completed.
+func maybeCompactBackups(
+	ctx context.Context,
+	execCtx sql.JobExecContext,
+	jobDetails jobspb.BackupDetails,
+	kmsEnv cloud.KMSEnv,
+) error {
+	threshold := backupCompactionThreshold.Get(&execCtx.ExecCfg().Settings.SV)
+	if threshold == 0 || jobDetails.RevisionHistory || !jobDetails.Destination.Exists ||
+		(scheduleOnly && jobDetails.ScheduleID == 0) {
+		return nil
+	}
+	chain, _, _, _, err := getBackupChain(ctx, execCtx, jobDetails, kmsEnv)
+	if err != nil {
+		return err
+	}
+	if int64(len(chain)) < threshold {
+		return nil
+	}
+	compactStart, compactEnd, err := minimumSizeDeltaHeuristic(ctx, execCtx, chain)
+	if err != nil {
+		return err
+	}
+	startTS, endTS := chain[compactStart].StartTime, chain[compactEnd-1].EndTime
+	log.Infof(ctx, "compacting backups from %s to %s", startTS, endTS)
+	return execCtx.ExecCfg().InternalDB.Txn(ctx, func(ctx context.Context,
+		txn isql.Txn) error {
+		var encryptionBytes []byte
+		if jobDetails.EncryptionOptions != nil {
+			encryptionBytes, err = jobDetails.EncryptionOptions.Marshal()
+			if err != nil {
+				return err
+			}
+		}
+		_, err := txn.ExecEx(
+			ctx,
+			"start-compaction-job",
+			txn.KV(),
+			sessiondata.NoSessionDataOverride,
+			`SELECT crdb_internal.backup_compaction($1, $2, $3::BYTES, $4::DECIMAL, $5::DECIMAL)`,
```

Comment:
heh, so you decided to use the built in after all?

Reply:
Yeah I realized that `StartCompactionJob` takes in the `PlanHookState` from the builtin, but in this scheduled backup path, we have a `JobExecContext`. And I spent some time trying to figure out how I could consolidate this before realizing this was a futile effort and throwing in the towel.